src/adapters/bidderFactory.js

```diff
@@ -0,0 +1,240 @@
+        fillNoBids();
+        return;
+      }
+
+      let requests = spec.buildRequests(bidRequests);
+      if (!requests || requests.length === 0) {
+        fillNoBids();
+        return;
+      }
+      if (!Array.isArray(requests)) {
+        requests = [requests];
+      }
+
+      // Callbacks don't compose as nicely as Promises. We should call fillNoBids() once _all_ the
+      // Server requests have returned and been processed. Since `ajax` accepts a single callback,
+      // we need to rig up a function which only executes after all the requests have been responded.
+      const onResponse = delayExecution(fillNoBids, requests.length)
+      requests.forEach(processRequest);
+
+      function processRequest(request) {
+        switch (request.type) {
+          case GET:
+            ajax(
+              `${request.endpoint}?${parseQueryStringParameters(request.data)}`,
+              {
+                success: onSuccess,
+                error: onFailure
+              },
+              undefined,
+              {
+                method: GET,
+                withCredentials: true
+              }
+            );
+            break;
+          case POST:
+            ajax(
+              request.endpoint,
+              {
+                success: onSuccess,
+                error: onFailure
+              },
+              typeof request.data === 'string' ? request.data : JSON.stringify(request.data),
+              {
+                method: POST,
+                contentType: 'text/plain',
+                withCredentials: true
+              }
+            );
+            break;
+          default:
+            logWarn(`Skipping invalid request from ${spec.code}. Request type ${request.type} must be GET or POST`);
+            onResponse();
+        }
+      }
+
+      // If the server responds successfully, use the adapter code to unpack the Bids from it.
+      // If the adapter code fails, no bids should be added. After all the bids have been added, make
+      // sure to call the `onResponse` function so that we're one step closer to calling fillNoBids().
+      function onSuccess(response) {
+        const bidRequestMap = {};
+        bidRequests.forEach(bid => {
+          bidRequestMap[bid.bidId] = bid;
+        });
+        function addBidUsingRequestMap(bid) {
+          const bidRequest = bidRequestMap[bid.requestId];
+          if (bidRequest) {
+            const prebidBid = Object.assign(bidfactory.createBid(STATUS.GOOD, bidRequest), bid);
+            addBidWithCode(bidRequest.placementCode, prebidBid);
+          } else {
+            logWarn(`Bidder ${spec.code} made bid for unknown request ID: ${bid.requestId}. Ignoring.`);
+          }
+        }
+
+        let bids;
+        try {
+          bids = spec.interpretResponse(response);
+        } catch (err) {
+          logError(`Bidder ${spec.code} failed to interpret the server's response. Continuing without bids`, null, err);
+          onResponse();
```

Comment:
I think the calling of `onResponse` here is a good use case for a `finally` clause so you don't have to scatter as many calls around.

I'd wrap all this code in a try block and put the `onResponse` in a finally so it's obvious that it will always be called.

Reply:
I like the goal here... but I'm really trying not to catch unexpected errors either.

For example, while writing this I originally had a bug in my `addBidUsingRequestMap` function where something was unexpectedly undefined.

I easily may never have caught that if it'd been inside the `try` block.